Adding feature for banks to broadcast announcements to partners issue: 3390

cotabas opened the pull request:
<!--Read comments, before committing pull request read checklist again

# Checklist:

- I have performed a self-review of my own code,
- I have commented my code, particularly in hard-to-understand areas,
- I have made corresponding changes to the documentation,
- I have added tests that prove my fix is effective or that my feature works,
- New and existing unit tests pass locally with my changes ("bundle exec rake"),
- Title include "WIP" if work is in progress.

-->

Resolves #3390 <!--fill issue number-->

### Description
New feature added to allow banks to broadcast announcements to the partners. 

For the banks there's a link on their side bar that goes to a table of the current announcements where they edit or delete them. There's a button at the top for creating new announcements (screen shots below). 

Each announcement consists of a message, with a 500 character limit, an optional link for more info, and an optional expiration date. The creating user and organization are also stored so this table can be used for admin to bank announcements.

Expired announcements persist on the bank side with a red 'expired' tag until deleted but are not shown to partners.

The partner dash board has a card at the top that displays the announcements, or no card if there's no announcement


<!-- Please include a summary of the change and which issue is fixed. 
Please also include relevant motivation and context.
Guide questions:
  - What motivated this change (if not already described in an issue)?
  - What alternative solutions did you consider?
  - What are the tradeoffs for your solution?
   
List any dependencies that are required for this change. (gems, js libraries, etc.)

Include anything else we should know about. -->

### Type of change

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)
* This change requires a documentation update



### How Has This Been Tested?
There are tests for the model and how it validates entries and CRUD tests for the controller

All tests pass, new and old
<!-- Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Do we need to do anything else to verify your changes? 
If so, provide instructions (including any relevant configuration) so that we can reproduce? -->

### Screenshots

Main bank page, I highlighted the sidebar menu change
![ba_bank_side](https://user-images.githubusercontent.com/94578504/223336328-30a2ad43-b938-47e1-8684-8ee372e68f24.png)
New announcement bank page, edit looks the same except for the title and breadcrumbs
![ba_bank_new](https://user-images.githubusercontent.com/94578504/223336333-a33339c9-13ca-4749-8468-30e1119c9939.png)
Partner dashboard, the card is only shown if there's an announcement and it only shows bank announcements that haven't expired

One announcement:
![image](https://user-images.githubusercontent.com/94578504/223350128-71cb73c9-17ba-4415-aa37-e3a156cefe0f.png)
Multiple announcements:
![ba_partner_dash](https://user-images.githubusercontent.com/94578504/223336327-e2913c0b-c97e-4e45-a934-6dbc74a01b5f.png)


<!--Optional. Delete if not relevant. 
Include screenshots (before / after) for style changes, highlight edited element.-->

cotabas commented:
I'm new, so hopefully I didn't do anything too weird. I'm happy to change anything and greatly appreciate any feedback

dorner commented on app/controllers/broadcast_announcements_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  before_action :set_broadcast_announcement, only: %i[edit update destroy]
+
+  def index
+    @broadcast_announcements = BroadcastAnnouncement.all.select { |announcement| announcement.organization_id == current_organization.id }
```
This should be a `where` query rather than `select`, which would load every announcement in the database into memory and then filter them after the fact.

dorner commented on app/controllers/broadcast_announcements_controller.rb:
```diff
@@ -0,0 +1,54 @@
+      if @broadcast_announcement.save
+        format.html { redirect_to broadcast_announcements_url, notice: "Broadcast announcement was successfully created." }
+      else
+        format.html { render :new, status: :unprocessable_entity }
```
Are we swallowing the errors? They should be displayed to the user somehow (e.g. via an `alert` flash message).

cotabas replied:
I think I might be misunderstanding you, is there another error that wouldn't show up?
Right now the form errors reload the :new page and show up like this:
![image](https://user-images.githubusercontent.com/94578504/223864018-4c544559-51ec-485b-82fd-c27b29594cac.png)

I fixed most of the things, I'll have more time this weekend to finish with this and the specs

dorner replied:
My bad, looks good here :)

dorner commented on app/controllers/partners/dashboards_controller.rb:
```diff
@@ -24,6 +24,11 @@ def show
+
+      @broadcast_announcements = BroadcastAnnouncement.all.select { |announcement|
+        Role.find(announcement.user_id).name == 'org_user' &&
+          (announcement.expiry.nil? || announcement.expiry.strftime('%y %m %d') >= DateTime.now.strftime('%y %m %d'))
```
A couple of issues here:

*   Again, it's loading the whole announcement table into memory - the conditions should be part of a `where` clause.
*   This checks that the announcer is an `org_user` but not that the announcer is actually the owning organization.

dorner commented on app/models/broadcast_announcement.rb:
```diff
@@ -0,0 +1,19 @@
+#  message         :text
+#  created_at      :datetime         not null
+#  updated_at      :datetime         not null
+#  organization_id :bigint           not null
```
We should allow nulls here so that in the future we can have admin announcements.

dorner commented on app/views/broadcast_announcements/_broadcast_announcement.html.erb:
```diff
@@ -0,0 +1,14 @@
+<tr id="<%= dom_id broadcast_announcement %>">
+  <td><%= broadcast_announcement.message %></td>
+  <td><%= link_to  broadcast_announcement.link, broadcast_announcement.link %></td>
+  <td><%= User.find(broadcast_announcement.user_id).name %></td>
```
You should be able to write `broadcast_announcement.user.name` here. Or if you want to be a bit more defensive, `broadcast_announcement.user&.name || "N/A"`.

dorner commented on app/views/broadcast_announcements/_broadcast_announcement.html.erb:
```diff
@@ -0,0 +1,14 @@
+  <td><%= broadcast_announcement.expiry %>
+    <% unless broadcast_announcement.expiry.nil? || broadcast_announcement.expiry.strftime('%y %m %d') >= DateTime.now.strftime('%y %m %d') %>
+      <span class="badge badge-danger">Expired</span>
+    <% end %>
```
This should be a method called \`expired?\` on the `BroadcastAnnouncement` model.

dorner commented on app/views/broadcast_announcements/_form.html.erb:
```diff
@@ -0,0 +1,41 @@
+              <div class="card-body w-50">
+                <%= f.hidden_field :user_id, value: current_user.id %>
+                <%= f.hidden_field :organization_id, value: current_organization.id %>
+                <%= f.label :message, "Announcement " %><i>(500 character max)</i>
```
Needs a space here between the label and the `<i>` tag.

dorner commented on app/views/broadcast_announcements/_form.html.erb:
```diff
@@ -0,0 +1,41 @@
+                        select_input.forEach((input) => {input.disabled=true});
+                        }" >
+                <label for="expiry_check">Expiration date</label>
+                <%= f.input :expiry, label: false, disabled: true, class: "form-control" %>
```
You should be using a calendar control here - then you don't need the checkbox. You can use Litepicker for this: https://github.com/rubyforgood/human-essentials/blob/2c7bd3d70f0e7f46a31e9d6e1872f2fa195de1ca/app/javascript/application.js#L88

dorner commented on app/views/partners/dashboards/show.html.erb:
```diff
@@ -1,4 +1,5 @@
 <header class="container mx-auto mb-5">
+  <%= render 'broadcast_announcement' unless @broadcast_announcements.empty? %>
```
Use `if x.any?` instead of `unless x.empty?` - positive conditions are always easier to read.

dorner commented on spec/requests/broadcast_announcements_spec.rb:
```diff
@@ -0,0 +1,20 @@
+    @test_a.destroy
+    expect(BroadcastAnnouncement.count).to eq(0)
+  end
+end
```
There need to be request specs as well as a check on the dashboard spec if an announcement exists that it shows up.

dorner reviewed (changes requested):
@cotabas this is a great first pass! I added some suggestions.

cotabas commented:
@dorner Thank you so much for taking the time to go through my code a provide feedback

cielf commented:
Excited to see this one in progress, @cotabas!  The essentials banks are going to be pretty happy to get this.

cielf reviewed (changes requested):
Hi @cotabas.    I really like what I see, and I'll leave the technical aspects to @dorner,  but I'm going to ask for some presentation tweaks to the announcements on the partner dashboard:
    
1/  If I'm reading this right,  the announcements are presented to the partners in chronological order.   It should be reverse chronological order - so the newest news is on top, and the older announcements that the partner has probably already seen are on the bottom.   

2/  The partners each work with a single bank,  so we can move the bank name up into the announcements header.  

3/  Please include the year in the announcement date -- I'm thinking about what it will be like in a year, when we have no-expiry announcements from April 2023 and new announcements from March 2024.  (If you want to get fancy, having the current year suppressed, and the date showing for earlier years would be fine, but we don't need it)

Thanks again for your work on this!

cotabas commented:
Hey @cielf, thanks for reviewing, I've enjoyed working on this feature and I'm happy to hear it will be helpful. 
I finished the changes @dorner requested.
I've also made all the changes you've requested, I did go the fancy route and have it only displaying the full date if it's from a previous year. 
Please let me know if there's anything else I can change

dorner commented on spec/models/broadcast_announcement_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  it "expired? should be true if the announcement is expired" do
+    test_a = build(:broadcast_announcement, expiry: 2.days.ago)
+    expect(test_a.expired?).to eq(true)
```
There should be a check for false as well :)

cotabas replied:
gotcha

dorner commented on spec/requests/broadcast_announcements_spec.rb:
```diff
@@ -0,0 +1,124 @@
+    it "redirects after delete" do
+      announcement = BroadcastAnnouncement.create! valid_attributes
+      delete broadcast_announcement_url(announcement)
+      expect(response).to have_http_status(:redirect)
```
It's OK to have multiple expectations on the same test especially for request / system tests, which take a long time to run. If you're running the exact same code, just put your expectations together.

cotabas replied:
That makes sense. Thank you again for taking the time to explain all this to me

cielf reviewed (commented):
I see the changes I asked for.   The one thing I am uncertain about is if we should perhaps have a test for the ordering of the announcements on the partner dashboard.  @dorner - what do you think?  There's a balance between "it's expected behaviour" and "system tests are slow and expensive"

dorner commented:
I think we can balance this by adding a method on the announcement class that returns the list of announcements to display given an organization ID - we can unit test that method and ensure that it's the one called in the request test.

cotabas commented:
I added the announcement method as suggested but I'm having trouble testing it. When I try to create an announcement in the spec with a different organization_id, or no organization_id, it fails with Validation failed: Organization must exist.
I removed the not null on organization_id as suggested earlier so I'm not sure what's going on.
I've updated my test database since that migration

```
ActiveRecord::RecordInvalid:
       Validation failed: Organization must exist
```

cotabas commented:
Thank you @dorner apparently I've never used optional references
Let me know if I need to change anything else

dorner commented on app/models/broadcast_announcement.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def self.filter_announcements(parent_org)
+    BroadcastAnnouncement.where(organization_id: parent_org)
```
To include "no-org" (admin) announcements, this should be `where(organization_id: [parent_org, nil].uniq)`.

cielf replied:
Hmm.   [This comment is largely an aside to @dorner]  I know this work is going to be the basis for the next step -- which is the admin announcements -- but/and the initial model only has   admin announcements going down to the banks.    I'm not saying don't make this change, but that we may have to think about a broadcast level when we implement the next step.

cotabas replied:
Ok, you do want the future admin announcements to be displayed on the partner dashboard? I was thinking the future admin announcements would only be for the banks

cielf replied:
There's a case for admins being able to communicate to the partners, but it would be only some announcements .   I think it's out of scope for this issue. @dorner?  Are you ok with that?

dorner replied:
Ah OK - yeah we can leave it as is then and address the admin stuff later.

dorner reviewed (changes requested):
One more comment and I think we should be good after that!

cielf commented:
@cotabas Just released this one too!
